### ISSUE TITLE
dotnet-watch: Allow multiple dotnet commands 

### ISSUE THREAD

**Issue by arexsutton** [aspnet/DotNetTools#246]
I have been loving `dotnet watch`. Thank you for all your hard work. Also, sup @natemcmaster!

So I have a use case where I would like to run both `dotnet watch test` and `dotnet watch run` in sequence. Ideally, I would be able to say something like `dotnet watch test run` and have it:

```
while true
  compile
  execute dotnet test
  execute dotnet run
  watch for file changes
```

Currently, I do this:
```
#!/bin/bash
dotnet watch test -explore:cases -noresult &
trap "trap - SIGTERM && kill -- -$$" SIGINT SIGTERM EXIT
dotnet watch run
```

However, this approach is a bit nasty, and it exposes one problem some of us have seen with the `.IncrementalCache` file. Given two compilations are trying to read/modify `.IncrementalCache` in parallel, it increases the frequency of the following error:

```
Could not write the incremental cache file: 
/home/username/project/obj/Debug/netcoreapp1.0/.IncrementalCache
```

Note: A partial workaround is `ln -s /dev/null obj/Debug/netcoreapp1.0/.IncrementalCache`, at least then we don't have to restart the watcher process every time the compilation crashes due to this error.

Anyway, I read that `.IncrementalCache` is going away, so this aspect of the issue may not even be relevant in the near future. However, I think the request is potentially still good.

Given the 1-dimensional nature of command line arguments and dynamic nature of which dotnet commands are available, this is potentially a Hard UI Problem:tm:. Consider the command line arguments I like to pass: `dotnet test -explore:cases -noresult`... does that mean that our command looks like this `dotnet watch test -explore:cases -noresult run`? 

My understanding is that we can't exactly be in the business of knowing what dotnet commands are available. And what if certain commands expect non-dashed option arguments, like `dotnet operation typename`, then we can't logically deduce what is and is not an executable dotnet command. Here is the current behavior there:

```
$ dotnet watch foo
[DotNetWatcher] info: Running dotnet with the following arguments: foo
[DotNetWatcher] info: dotnet process id: 14470
No executable found matching command "dotnet-foo"

$dotnet watch foo bar
[DotNetWatcher] info: Running dotnet with the following arguments: foo bar
[DotNetWatcher] info: dotnet process id: 14628
No executable found matching command "dotnet-foo"
```

A note from `dotnet watch --help`:
```
  The special option '--' is used to delimit the end of the options and
  the beginning of arguments that will be passed to the child dotnet process.
  Its use is optional. When the special option '--' is not used,
  dotnet-watch will use the first unrecognized argument as the beginning
  of all arguments passed into the child dotnet process.
```

Would it make sense to overload the usage of this special argument to delimit multiple subcommands like so?

```
dotnet watch --someOption -- test -explore:cases -noresult -- run
and
dotnet watch test -explore:cases -noresult -- run

```

There're my thoughts, let me know what you think. Anyone else have this use case? Anyone have a better idea on how to solve this use case?

Thanks for your time,
Anthony Rex Sutton

**Comment by natemcmaster** [comment]
Hey @arexsutton! Nice to see a familiar github handle :)

Creating usable and understandable command-line syntax has been the most difficult challenge of dotnet-watch. My initial reaction here is that supporting multiple dotnet commands might complicate it too much, but open to suggestions.

Also, some of this complexity may be solved when you upgrade from project.json to MSBuild. For example, you could run multiple MSBuild targets:

   dotnet watch msbuild "/t:Test;Run"

Side note: I find it curious that you want both `dotnet run` and `dotnet test`. I haven't seen a test project and a console app combined into one project.

**Comment by arexsutton** [comment]
So I changed my project to be two projects (and thus compiled DLLs), one with the core code and the second being the test code. It seems the same issue does persist, seemingly as a result of the core project being built with the `.IncrementalCache` file located in the core project's target directory, even when being built from the test project. Therefore running dotnet watch test on the test project and dotnet watch run on the core project still causes colliding on shared resources.

I'm interested that you consider MSBuild to be an upgrade from the project.json method. It's been my perception (which could easily be wrong) that MSBuild is more or less dying if not already dead, for new solutions. In any case, I'm very much loving the project.json way and do not like the MSBuild way, so I would be disinclined to switch. If that's my only solution at the end of the day, I'll probably just deal with this problem with hacks and workarounds instead, and hope `.IncrementalCache` goes away sooner rather than later.

**Comment by natemcmaster** [comment]
> Therefore running dotnet watch test on the test project and dotnet watch run on the core project still causes colliding on shared resources.

You should be able to run just one dotnet-watch process on the top-level project. dotnet-watch was designed to watch a project and all of the projects it depends on.

> MSBuild is more or less dying if not already dead

Actually, project.json is about to be killed off. See https://blogs.msdn.microsoft.com/dotnet/2016/05/23/changes-to-project-json/. VS2017 does not support project.json. dotnet-watch is also deprecating support for this. The 'preview4' version is the last one that will support project.json. All efforts going forward are to support MSBuild.

**Comment by muratg** [comment]
Assigning to @natemcmaster for investigating possible workarounds to this.

**Comment by natemcmaster** [comment]
@arexsutton we recently released the 1.0.0 version of dotnet-watch which means we have dropped support for project.json. For 1.0, the 1.0 command-line syntax does not allow multiple commands. I think it is unlikely we will add this in the future.

That said, MSbuild gives you the flexibility to program your own commands. For example, it is possible to create an msbuild file that will allow you to watch multiple projects and run multiple commands. I'm going to [add samples](https://github.com/aspnet/DotNetTools/issues/274) to our repo to show how to do this, but here is the gist of how to make this work.

Setup:
```
+ watch.proj
- test/
     + test.csproj
     + UnitTest.cs
- web/
     + web.csproj
     + Startup.cs
```

```xml
<!-- contents of watch.proj -->
<Project>
  <PropertyGroup>
    <TargetFramework>netcoreapp1.0</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <DotNetCliToolReference Include="Microsoft.DotNet.Watcher.Tools" Version="1.0.0" />
    <ProjectReference Include="web\web.csproj" />
    <ProjectReference Include="test\test.csproj" />
  </ItemGroup>

  <Target Name="TestAndRun">
    <Exec Command="dotnet test" WorkingDirectory="test/" />
    <Exec Command="dotnet run" WorkingDirectory="web/" />
  </Target>

  <Import Project="$(MSBuildToolsPath)\Microsoft.Common.targets"/>
</Project>
```

Command line:
from the folder containing `watch.proj`
```
dotnet restore watch.proj
dotnet watch /t:TestAndRun
```

Feel free to continue discussing. Closing for now as I believe running multiple commands is already possible.

**Comment by alohaninja** [comment]
This seems to work if your commands run in sequence, but you can't run parallel projects. Say for example you have a microservice structure or multiple service tiers - an **Web, Middleware,** and **Data** tier projects all running on kestrel. How do you build this out so that `dotnet watch` doesn't stomp over itself? It would be nice to have support for something like this...

```
<!-- contents of watch.proj -->
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netcoreapp1.0</TargetFramework>
  </PropertyGroup>
  <ItemGroup>
    <DotNetCliToolReference Include="Microsoft.DotNet.Watcher.Tools" Version="1.0.0" />
    <ProjectReference Include="Web\Web.csproj" />
    <ProjectReference Include="Middleware\Middleware.csproj" />
    <ProjectReference Include="Data\Data.csproj" />
  </ItemGroup>
  <Target Name="RunApp">
    <Exec Command="dotnet run" WorkingDirectory="Data/" />
    <Exec Command="dotnet run" WorkingDirectory="Middleware/" />
    <Exec Command="dotnet run" WorkingDirectory="Web/" />
  </Target>
</Project>
```

**Comment by arexsutton** [comment]
@alohaninja Hey Aaron, I don't know if you're running `dotnet watch` from bash or a shell with job control, but assuming you are, this kind of approach has worked great for me (I run dotnet core on Linux, I haven't tested this approach on the Linux Subsystem for Windows):

```
#!/bin/bash

# Run watcher on unit test projects
pushd Web.Tests.Unit/
    dotnet restore
    dotnet watch test &
popd
pushd Middleware.Tests.Unit/
    dotnet restore
    dotnet watch test &
popd
pushd Data.Tests.Unit/
    dotnet restore
    dotnet watch test &
popd

# Run watcher for all my webservices
pushd Web/
    dotnet restore
    dotnet watch run &
popd
pushd Middleware/
    dotnet restore
    dotnet watch run &
popd
pushd Data/
    dotnet restore
    
    # When the script is killed e.g. with CTRL-C, make sure all spawned subprocesses die too
    trap "trap - SIGTERM && kill -- -$$" SIGINT SIGTERM EXIT

    # make sure to start the last subprocess in the foreground
    dotnet watch run 
popd
```
